Fix/inventories

Faithcaio opened the pull request:
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1848) | **SpongeCommon** | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/2232)

- Indexed access in base Inventory
- Redo lenses based on OrderedInventoryLens
- toString for lenses to help debugging
- remove observable lenses
- tracking opened ViewableInventories

Cybermaxke commented on src/main/java/org/spongepowered/common/item/inventory/InventoryTransactionResultImpl.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+
+    public InventoryTransactionResult and(InventoryTransactionResult other) {
```
Missing `@Override`s

Faithcaio committed:
Big Inventory API Update

Removed A LOT of unneeded generics
Tracking opened inventories
Remove usage of empty Optional ItemStack for empty stacks
